eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java

```diff
@@ -0,0 +1,131 @@
+    private static final String SERVICE_CLIENT_BUILDER = "ServiceClientBuilder";
+    private static final String BUILD_CLIENT = "buildClient";
+    private static final String BUILD_ASYNC_CLIENT = "buildAsyncClient";
+
+    private static boolean hasServiceClientBuilderAnnotation;
+    private static boolean hasAsyncClientBuilder;
+    private static boolean hasClientBuilder;
+
+    @Override
+    public void beginTree(DetailAST root) {
+        hasServiceClientBuilderAnnotation = false;
+        hasAsyncClientBuilder = false;
+        hasClientBuilder = false;
+    }
+
+    @Override
+    public void finishTree(DetailAST root) {
+        if (!hasServiceClientBuilderAnnotation) {
+            return;
+        }
+
+        if (!hasAsyncClientBuilder) {
+            log(root, String.format("The class annotated with @ServiceClientBuilder requires an asynchronous method, ''%s''", BUILD_ASYNC_CLIENT));
+        }
+        if (!hasClientBuilder) {
+            log(root, String.format("The class annotated with @ServiceClientBuilder requires a synchronous method, ''%s''", BUILD_CLIENT));
+        }
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                final DetailAST serviceClientAnnotationToken = getServiceClientAnnotation(token);
+                final String className = token.findFirstToken(TokenTypes.IDENT).getText();
+
+                hasServiceClientBuilderAnnotation = serviceClientAnnotationToken != null;
+
+                if (hasServiceClientBuilderAnnotation) {
+                    // Don't need to check if the 'serviceClients' exist. It is required when using @ServiceClientBuilder
+
+                    // HAS @ServiceClientBuilder annotation but NOT named the class <ServiceName>ClientBuilder
+                    if (!className.endsWith("ClientBuilder")) {
+                        log(token, String.format("Class annotated with @ServiceClientBuilder ''%s'' should be named <ServiceName>ClientBuilder.", className));
+                    }
+                } else {
+                    // No @ServiceClientBuilder annotation but HAS named the class <ServiceName>ClientBuilder
+                    if (className.endsWith("ClientBuilder")) {
+                        log(token, String.format("Class ''%s'' should be annotated with @ServiceClientBuilder.", className));
+                    }
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (!hasServiceClientBuilderAnnotation) {
```

Comment:
Is the order of `visitToken` call as per the array returned by `getRequiredTokens()`? If not, visitToken might be called with `METHOD_DEF` first and then `CLASS_DEF`. If this happens, `hasServiceClientBuilderAnnotation` would not be updated yet and will be `false` resulting in skipping method validation.

Reply:
Usually, we will put the inner nested class at the end of class def.  But it could still have a scenario that an inner nested class before methods in the same class. I modified the code to support the scenario in the new commit soon.